### ISSUE TITLE
Text search

### ISSUE THREAD

**Issue by jenniferplusplus** [Letterbook/Letterbook#23]

**Comment by jenniferplusplus** [comment]
I was thinking earlier about ways to constrain and partition search indexes that might help, and I want to make a note of them before I forget. For the most part, I don't think these options are exclusive of each other. That is to say, it would be possible to index posts that are both public and local only, and search could be restricted to the author's own posts only, all at the same time.

- Self only. That is, an account (profile?) could search their own posts, but not others.
- Public only. I think this should be true regardless. No indexing unlisted, followers-only, or private posts.
- Mentioned only. Search restricted to posts that mention the searching account.
- Interacted only. Search restricted to posts the searching account has interacted with (like, boost, reply, bookmark). This seems hard, at a technical level. Seems like it would require a dedicated index per-account.
- Local only. This addresses some cost and difficulty concerns. Definitely makes search less useful, though.
- Hashtags. I think this is just necessary for feature parity with mastodon.
- Top level only. Don't index replies, only the original post.

In all cases, Letterbook should respect any signal we can find that a post or actor shouldn't be made searchable. And we super especially should not crawl any social or interaction graphs or perform universal indexing on all federated content we receive.

**Comment by runewake2** [comment]
I think the difficult part here is that Mastodon does not support search right now. That means a lot of people using the tool do not expect their content to be indexed. Since we are playing in that same room I do think we ought to respect that as much as possible. Potentially that could mean only indexing your own posts/bookmarks/interactions or searching hashtags.

Your work on #37 does create the potential for us to have additional internal metadata for local instance accounts about if that account has opted in to instance wide indexing. We might be able to do something around that as well.

Some of the abuse patterns for the search feature include things like watching keywords via bots/accounts in order to find and harass users. If there's a way we can minimize that in our implementation it might be received more warmly.

I do not think we should index users that have not opted in and I'm not sure a public post is enthusiastic consent to that opt-in question considering how the broader ecosystem handles this right now (no search, except over hashtags).

A potential option that I'm less sure about would be indexing followers-only where the search feature only returns results which match a user who you have been allowed to follow (or yourself). In that case consent was already given that you could follow that users posts. I think implementing this might be expensive and complex though.

**Comment by jenniferplusplus** [comment]
I agree that universal full text search isn't what most people signed up for. That's not what I want to do, either. But text search is more common than people make it out to be. Calckey does it. So does Wildebeast. And Mastodon if you give it an elastic db. It's just that not everyone does, because it's optional. Then there's things like tootfinder. And it's impossible to imagine that Threads won't be indexing everything. Which is to say that the genie is already most of the way out of the bottle. It would be ironic and terrible if the best way to search Mastodon was to use Threads.

I think search can be great for finding and building community, as long as we get the boundaries right with it.

**Comment by runewake2** [comment]
So how do we do search in a way that minimizes harm? I understand this might be a question that's more complex than we can answer ourselves.

I do think, used responsibly, that search is an important feature - especially with how meta some discussions on the federated platforms can get - without search it can be hard to break out of the social bubble of who you've already encountered.

**Comment by jenniferplusplus** [comment]
I'm not sure exactly. I have some ideas. Limiting indexing and access to those indexes is a lot of it. I think expiring things out of those indexes would also be a factor.
